pgx/src/spi.rs

```diff
@@ -494,7 +457,10 @@ impl SpiTupleTable {
                         let datum =
                             pg_sys::SPI_getbinval(heap_tuple, tupdesc, ordinal, &mut is_null);
```

Comment:
This call and the below `T::from_datum()` call should happen like so:

```rust
let datum = PgMemoryContexts::For(PgMemoryContexts::CurrentMemoryContext.parent()).switch_to(|context| {
   
      pg_sys::SPI_getbinval(heap_tuple, tupdesc, ordinal, &mut is_null);
      T::from_datum(datum, is_null)
});
```

We need Postgres to do the `SPI_getbinval` call in the parent context as well as the from_datum call.  This way, if the Datum doesn't need to be de-toasted but still allocated, it'll be in the right place.

Reply:
So, I tried to write it like this:

```rust
PgMemoryContexts::CurrentMemoryContext
                            .parent()
                            .expect("parent memory context is absent")
                            .switch_to(|_context| {
                                let mut is_null = false;
                                let datum = pg_sys::SPI_getbinval(
                                    heap_tuple,
                                    tupdesc,
                                    ordinal,
                                    &mut is_null,
                                );
                                T::from_polymorphic_datum(
                                    datum,
                                    is_null,
                                    pg_sys::SPI_gettypeid(tupdesc, ordinal),
                                )
                            })
```

The diff is:

```diff
diff --git a/pgx/src/spi.rs b/pgx/src/spi.rs
index c185d92e..262ff50c 100644
--- a/pgx/src/spi.rs
+++ b/pgx/src/spi.rs
@@ -453,18 +453,24 @@ impl SpiTupleTable {
                     } else {
                         let heap_tuple = std::slice::from_raw_parts((*self.table).vals, self.size)
                             [self.current as usize];
-                        let mut is_null = false;
-                        let datum =
-                            pg_sys::SPI_getbinval(heap_tuple, tupdesc, ordinal, &mut is_null);
-
-                        T::from_datum_in_memory_context(
-                            PgMemoryContexts::CurrentMemoryContext
-                                .parent()
-                                .expect("parent memory context is absent"),
-                            datum,
-                            is_null,
-                            pg_sys::SPI_gettypeid(tupdesc, ordinal),
-                        )
+
+                        PgMemoryContexts::CurrentMemoryContext
+                            .parent()
+                            .expect("parent memory context is absent")
+                            .switch_to(|_context| {
+                                let mut is_null = false;
+                                let datum = pg_sys::SPI_getbinval(
+                                    heap_tuple,
+                                    tupdesc,
+                                    ordinal,
+                                    &mut is_null,
+                                );
+                                T::from_polymorphic_datum(
+                                    datum,
+                                    is_null,
+                                    pg_sys::SPI_gettypeid(tupdesc, ordinal),
+                                )
+                            })
                     }
                 },
                 None => panic!("TupDesc is NULL"),
```

and immediately, tests started failing, showing random data all over the place.

I've looked up [SPI_getbinval](https://www.postgresql.org/docs/current/spi-spi-getbinval.html) documentation, and it states: "This function does not allocate new space for the datum. In the case of a pass-by-reference data type, the return value will be a pointer into the passed row."


Could it be the reason why my code worked fine and the above didn't? I don't necessarily assume that the above code is what you intended me to try. Maybe I didn't get what you wanted me to do.

My reading suggests that calling `SPI_getbinval` is the parent context is wrong if it returns pointers (as states above).

Reply:
@eeeebbbbrrrr any thoughts on this?

Reply:
I really hate SPI.

After reading the SPI_getbinval docs myself, I don't think it's that calling SPI_getbinval in any context is wrong per se.  I think what it returns points to data that was already allocated for us in CurrentMemoryContext, which is the one created by SPI_connect.

I should have read the docs before my comment here.  I'm sorry.

I now think what you had before using `FromDatum::from_datum_in_memory_context()` is the correct thing to do, as that function is supposed to **copy** the datum into the provided context, which just switching memory contexts like this doesn't actually do.

So yeah, going back to what you originally had is correct.  But now it leaves me with a concern around "what if our current (and future!) `::from_datum_in_memory_context()` impls are bugged.

<short pause>

Geez, I think most of them are bugged right now by virtue of the fact the use the default implementation.

Let me think on this a bit more...

Reply:
> Geez, I think most of them are bugged right now by virtue of the fact they use the default implementation.

Well, wait.  No, I think they're okay.  We only have non-default impls for (rust) reference types like `&str` and `&[u8]` (and a few others).

For a minute I was second-guessing the code and thinking it'd be broken for `String` or `Vec<u8>` since those aren't first copying the Datum, but in fact they are -- straight into Rust-allocated memory, which is totally fine.

So yeah, lets go back to your original code and I think we're good.  Thanks for your tenacity!

Reply:
I didn't commit the suggested update, so it's at what I wrote initially. Anything I can do to  make this mergeable?